shared/protocol/src/main/java/io/pravega/shared/protocol/netty/AppendDecoder.java

```diff
@@ -95,29 +95,35 @@ public Request processCommand(WireCommand command) throws Exception {
                     ca.getExpectedOffset(), ca.getRequestId());
             break;
         case APPEND_BLOCK:
-            getSegment(((WireCommands.AppendBlock) command).getWriterId());
             currentBlock = (WireCommands.AppendBlock) command;
             result = null;
             break;
         case APPEND_BLOCK_END:
             WireCommands.AppendBlockEnd blockEnd = (WireCommands.AppendBlockEnd) command;
-            if (currentBlock == null) {
-                log.warn("Received AppendBlockEnd {} without AppendBlock", blockEnd);
-                throw new InvalidMessageException("AppendBlockEnd without AppendBlock.");
-            }
             UUID writerId = blockEnd.getWriterId();
-            if (!writerId.equals(currentBlock.getWriterId())) {
-                throw new InvalidMessageException("AppendBlockEnd for wrong connection.");
-            }
             segment = getSegment(writerId);
+            int sizeOfWholeEventsInBlock = blockEnd.getSizeOfWholeEvents();
+            ByteBuf appendDataBuf;
+            if (blockEnd.numEvents <= 0) {
+                throw new InvalidMessageException("Invalid number of events in block.");
+            }
             if (blockEnd.getLastEventNumber() < segment.lastEventNumber) {
                 throw new InvalidMessageException("Last event number went backwards.");
             }
-            int sizeOfWholeEventsInBlock = blockEnd.getSizeOfWholeEvents();
-            if (sizeOfWholeEventsInBlock > currentBlock.getData().readableBytes() || sizeOfWholeEventsInBlock < 0) {
-                throw new InvalidMessageException("Invalid SizeOfWholeEvents in block");
+            if (currentBlock != null) {
+               if (!currentBlock.getWriterId().equals(writerId)) {
+                   throw new InvalidMessageException("Writer ID mismatch between Append Block and Append block End.");
+               }
+               if (sizeOfWholeEventsInBlock > currentBlock.getData().readableBytes() || sizeOfWholeEventsInBlock < 0) {
+                    throw new InvalidMessageException("Invalid SizeOfWholeEvents in block");
```

Comment:
Include the value in the exception. It will help to debug.

Reply:
yes its included now.